# 10X Genomics Data Analysis (Seurat)
Course materials for Day 3 10X Genomics data analysis (Seurat) of the Single-Cell RNAseq Training Course 2021.

## Session Overview
In this session we will cover:
1. The theory behind ‘downstream’ single-cell analysis with a focus on 10X Genomics data
2. How to analyse a single-cell dataset using the Seurat Package
3. Address some of the assumptions/difficulties you may run into when analysing single-cell data

## Included in this Repo
* main course script: `ei_10X_genomics_analysis_pratical.Rmd`
* static web-version of the script: `ei_10X_genomics_analysis_pratical.html`
* `data` folder with the data for the course
  * `10X/filtered_gene_bc_matrices`: PBMC example data
  * `ss2`: PBMC smart-seq2 example data (not used during the session)
* `slides`: slides used alongside the hands on session
* `cellranger`: additional resources and information for using cellranger

## How to download the repo
1.To go this GitHub Repo: https://github.com/mattmadgwick/10XGenomicsDataAnalysis
2. Go over to the green code button on the right-hand side and then go to Download ZIP .
3. This should start the download of the data.
4. Once it has downloaded you will need to uncompress the folder and then you will have all the materials you need for the Hands on: 10X Genomics data analysis (Seurat) later today.

## How open the workbook in Rstudio
To open workbook:
1. Open Rstudio
2. On the top right hand corner click on the projects drop-down.
3. Click `Open Project`
4. Navigate the where you downloaded the course materials
5. Click on the folder 10XGenomicsDataAnalysis-main
6. Then click `Open`.


## Useful links
* Seurat packages website: https://satijalab.org/seurat/
* 10X Genomics Example datasets: https://www.10xgenomics.com/resources/datasets
* Cellragner: https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger
* R tutorials: https://www.codecademy.com/learn/learn-r
